File: src/test/java/com/github/natche/gravatarjavaclient/profile/GravatarProfileTokenProviderTest.java
package com.github.natche.gravatarjavaclient.profile;

import org.junit.jupiter.api.Test;

import java.nio.charset.StandardCharsets;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for {@link GravatarProfileTokenProvider}s.
 */
public class GravatarProfileTokenProviderTest {
    /**
     * Creates a new instance for testing purposes.
     */
    GravatarProfileTokenProviderTest() {}

    /**
     *
     */
    @Test
    void testCreation() {
        assertThrows(NullPointerException.class,
                () -> new GravatarProfileTokenProvider(null, null));
        assertThrows(NullPointerException.class,
                () -> new GravatarProfileTokenProvider(() -> new byte[0], null));
        assertThrows(IllegalArgumentException.class,
                () -> new GravatarProfileTokenProvider(() -> new byte[0], ""));
        assertThrows(IllegalArgumentException.class,
                () -> new GravatarProfileTokenProvider(() -> new byte[0], "    "));
        assertDoesNotThrow(
                () -> new GravatarProfileTokenProvider(() -> new byte[0], "source"));
    }

    /**
     *
     */
    @Test
    void testGetToken() {
        byte[] bytes = "PostMalone".getBytes(StandardCharsets.US_ASCII);
        GravatarProfileTokenProvider provider = new GravatarProfileTokenProvider(() -> bytes, "source");
        assertEquals(bytes, provider.getToken());
    }

    /**
     *
     */
    @Test
    void testToString() {
        byte[] bytes = "PostMalone".getBytes(StandardCharsets.US_ASCII);
        GravatarProfileTokenProvider provider = new GravatarProfileTokenProvider(() -> bytes, "source");
        GravatarProfileTokenProvider nonEqualProvider
                = new GravatarProfileTokenProvider(() -> new byte[0], "another source");

        assertEquals("GravatarProfileTokenProvider{source=\"source\"}", provider.toString());
        assertEquals("GravatarProfileTokenProvider{source=\"another source\"}", nonEqualProvider.toString());
    }

    /**
     *
     */
    @Test
    void testEquals() {
        byte[] bytes = "PostMalone".getBytes(StandardCharsets.US_ASCII);
        GravatarProfileTokenProvider provider = new GravatarProfileTokenProvider(() -> bytes, "source");
        GravatarProfileTokenProvider equalProvider = new GravatarProfileTokenProvider(() -> bytes, "source");
        GravatarProfileTokenProvider nonEqualProvider
                = new GravatarProfileTokenProvider(() -> new byte[0], "another source");

        assertEquals(provider, provider);
        assertEquals(provider, equalProvider);
        assertNotEquals(provider, nonEqualProvider);
        assertNotEquals(provider, new Object());
    }

    /**
     *
     */
    @Test
    void testHashCode() {
        byte[] bytes = "PostMalone".getBytes(StandardCharsets.US_ASCII);
        GravatarProfileTokenProvider provider = new GravatarProfileTokenProvider(() -> bytes, "source");
        GravatarProfileTokenProvider equalProvider = new GravatarProfileTokenProvider(() -> bytes, "source");
        GravatarProfileTokenProvider nonEqualProvider
                = new GravatarProfileTokenProvider(() -> new byte[0], "another source");

        assertEquals(-896505829, provider.hashCode());
        assertEquals(-896505829, equalProvider.hashCode());
        assertEquals(1165062104, nonEqualProvider.hashCode());
        assertNotEquals(equalProvider.hashCode(), nonEqualProvider.hashCode());
    }
}

File: src/test/java/com/github/natche/gravatarjavaclient/profile/GravatarProfileTokenProviderTest.kt
package com.github.natche.gravatarjavaclient.profile

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.nio.charset.StandardCharsets

/**
 * Tests for [GravatarProfileTokenProvider]s.
 */
class GravatarProfileTokenProviderTest
/**
 * Creates a new instance for testing purposes.
 */
internal constructor() {
    /**
     * Tests for creation.
     */
    @Test
    fun testCreation() {
        Assertions.assertThrows(
            NullPointerException::class.java
        ) { GravatarProfileTokenProvider(null, null) }
        Assertions.assertThrows(
            NullPointerException::class.java
        ) { GravatarProfileTokenProvider({ ByteArray(0) }, null) }
        Assertions.assertThrows(
            IllegalArgumentException::class.java
        ) { GravatarProfileTokenProvider({ ByteArray(0) }, "") }
        Assertions.assertThrows(
            IllegalArgumentException::class.java
        ) { GravatarProfileTokenProvider({ ByteArray(0) }, "    ") }
        Assertions.assertDoesNotThrow<GravatarProfileTokenProvider> {
            GravatarProfileTokenProvider(
                { ByteArray(0) }, "source"
            )
        }
    }

    /**
     * Tests for the get token method.
     */
    @Test
    fun testGetToken() {
        val bytes = "PostMalone".toByteArray(StandardCharsets.US_ASCII)
        val provider = GravatarProfileTokenProvider({ bytes }, "source")
        Assertions.assertEquals(bytes, provider.token)
    }

    /**
     * Test for the to string method.
     */
    @Test
    fun testToString() {
        val bytes = "PostMalone".toByteArray(StandardCharsets.US_ASCII)
        val provider = GravatarProfileTokenProvider({ bytes }, "source")
        val nonEqualProvider = GravatarProfileTokenProvider({ ByteArray(0) }, "another source")
        Assertions.assertEquals("GravatarProfileTokenProvider{source=\"source\"}", provider.toString())
        Assertions.assertEquals("GravatarProfileTokenProvider{source=\"another source\"}",
            nonEqualProvider.toString())
    }

    /**
     * Tests for the equals method.
     */
    @Test
    fun testEquals() {
        val bytes = "PostMalone".toByteArray(StandardCharsets.US_ASCII)
        val provider = GravatarProfileTokenProvider({ bytes }, "source")
        val equalProvider = GravatarProfileTokenProvider({ bytes }, "source")
        val nonEqualProvider = GravatarProfileTokenProvider({ ByteArray(0) }, "another source")
        Assertions.assertEquals(provider, provider)
        Assertions.assertEquals(provider, equalProvider)
        Assertions.assertNotEquals(provider, nonEqualProvider)
        Assertions.assertNotEquals(provider, Any())
    }

    /**
     * Tests for the hashcode method.
     */
    @Test
    fun testHashCode() {
        val bytes = "PostMalone".toByteArray(StandardCharsets.US_ASCII)
        val provider = GravatarProfileTokenProvider({ bytes }, "source")
        val equalProvider = GravatarProfileTokenProvider({ bytes }, "source")
        val nonEqualProvider = GravatarProfileTokenProvider({ ByteArray(0) }, "another source")
        Assertions.assertEquals(-896505829, provider.hashCode())
        Assertions.assertEquals(-896505829, equalProvider.hashCode())
        Assertions.assertEquals(1165062104, nonEqualProvider.hashCode())
        Assertions.assertNotEquals(equalProvider.hashCode(), nonEqualProvider.hashCode())
    }
}